add doc for bgp autodiscovery (#39645)
cilium-team/cilium: Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst

          peerConfigRef:
            name: "cilium-peer"

Auto-Discovery
--------------

Cilium BGP Control Plane also supports automatic discovery of BGP peers.

When enabled, the auto-discovery feature self-configures the BGP peer's IP address automatically. Selection of the specific address is dependent on the ``mode`` enabled.

> Cilium BGP Control Plane currently supports ``DefaultGateway`` mode for auto-discovery under ``autoDiscovery`` field in ``CiliumBGPClusterConfig``

Reviewer: ```suggestion
Cilium BGP Control Plane currently supports ``DefaultGateway`` mode for auto-discovery under ``autoDiscovery`` field in ``CiliumBGPClusterConfig``.
```

Default Gateway Auto-Discovery
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The default gateway auto-discovery mode allows Cilium to automatically discover and establish BGP session with the default gateway (typically a Top-of-Rack switch) for a specified address family.

Reviewer: ```suggestion
The default gateway auto-discovery mode allows Cilium to automatically discover and establish BGP session with the default gateway (typically a Top-of-Rack (ToR) switch) for a specified address family.
```

To enable default gateway auto-discovery, configure the ``autoDiscovery`` field in the peer configuration:

.. code-block:: yaml

    peers:
    - name: "tor-switch"
      peerASN: 65000
      autoDiscovery:
        mode: "DefaultGateway"
        defaultGateway:
          addressFamily: ipv6  # Can be "ipv4" or "ipv6"
      peerConfigRef:
        name: "cilium-peer"

> Following are the ToR Switch BGP Configuration Requirements:

Reviewer: ```suggestion
Here are the ToR switch BGP configuration requirements:
```

* ToR switches must be configured with "bgp listen range" to support dynamic BGP neighbors. This configuration enables the ToR switch to accept BGP sessions from Cilium nodes by listening for connections from a specific IP prefix range, eliminating the need to know the exact peer address of each Cilium node.
  

>   FRR Reference: https://docs.frrouting.org/en/latest/bgp.html#clicmd-bgp-listen-range-A.B.C.D-M-X-X-X-X-M-peer-group-PGNAME

Reviewer: ```suggestion
  For more details, see the `FRR documentation <https://docs.frrouting.org/en/latest/bgp.html#clicmd-bgp-listen-range-A.B.C.D-M-X-X-X-X-M-peer-group-PGNAME>`__.
```

> * Configure each ToR Switch with the same local ASN (Autonomous System Number) to ensure Cilium configuration remains consistent across all cluster nodes.

Reviewer: ```suggestion
* Configure each ToR switch with the same local ASN (Autonomous System Number) to ensure Cilium configuration remains consistent across all cluster nodes.
```

> * To simply configuration using a single Cilium configuration across the cluster's nodes, each ToR switch should be configured using the same local ASN (Autonomous System Number).

Reviewer 2: Consider rewriting this line for clarity, I am not sure if you meant `simplify`.

Author: ya it was suppose to be `simplify`. i rewrote it for more clarity.

> .. code-block:: shell

Reviewer: `shell` will probably not render as you expect here, there's no prompt symbol.

```suggestion
For example::
```

  router bgp 65100
    neighbor CILIUM peer-group
    neighbor CILIUM local-as 65000 no-prepend replace-as

>     bgp listen range fd00:10:0:1::/64 peer-group CILIUM

Reviewer: Please introduce this snippet to tell what it's about. Does it refer to the two steps from the list above, or to the second only? If it refers to the second step only, it should be indented and made part of the second item, to avoid confusion.

Author: the example is for both the steps.
both the steps explain what the requirements are and then we provide an example for it so the reader will be able to understand that its related to the requirements?
If i provide more information about what the snippet is about, will it become redundant?

Reviewer: No more details needed, I simply meant a mention of what the snippet is about. The “For example” that you have now makes the transition, it tells it's an example and not some random configuration that should be applied in addition to other steps, for example. Thanks!

When configured with the above configuration, Cilium will:

* Determine the default gateway for the specified address family on each node
* Automatically establish a BGP session with the discovered gateway

> * Use the peer configuration referenced by ``peerConfigRef`` for session parameters

Reviewer: ```suggestion
Once this configuration is applied:

* Cilium determines the default gateway for the specified address family on each node
* It automatically establishes a BGP session with the discovered gateway
* It uses the peer configuration referenced by ``peerConfigRef`` for session parameters
```

> *WARNING: link-local address as default gateway is not supported.*

Reviewer: ```suggestion
.. warning::

   Link-local address as default gateway is not supported.
```

Multi-homing with Default Gateway Auto-Discovery
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In multi-homing setups, cilium node will be connected to two different Top-of-Rack switches. It will discover both the default gateways, but it will
choose the default route with the lower metric to establish the BGP session. It's important to note that it will create only one bgp session per address family
at a time. If the default route with the lower metric fails or gets changed, it will trigger a reconciliation to establish the BGP session with the default gateway

> of the other default route.

Reviewer: Present tense, please

```suggestion
In multi-homing setups, the Cilium node connects to two different Top-of-Rack switches. It discovers both the default gateways, but it will
picks the default route with the lower metric to establish the BGP session. It's important to note that Cilium creates only one BGP session per address family
at a time. A failure or a change of the default route with the lower metric triggers a reconciliation to establish the BGP session with the default gateway
of the other default route.
```

Example configuration:

.. code-block:: yaml

    bgpInstances:
    - name: "65001"
      localASN: 65001
      peers:
      - name: "instance-65001"
        peerASN: 65000
        autoDiscovery:
          mode: "DefaultGateway"
          defaultGateway:
            addressFamily: ipv6
        peerConfigRef:
          name: "cilium-peer"

Verification
~~~~~~~~~~~~

To verify that BGP sessions are established with the auto-discovered peers, use the ``cilium bgp peers`` command:

> .. code-block:: shell

Reviewer: ```suggestion
.. code-block:: shell-session
```

    $ cilium bgp peers
    Local AS   Peer AS   Peer Address         Session       Uptime   Family         Received   Advertised
    65001      65000     fd00:10:0:1::1:179   established   21m55s   ipv4/unicast   2          2    

>                                                                       ipv6/unicast   2          2

Reviewer: ```suggestion
    65001      65000     fd00:10:0:1::1:179   established   21m55s   ipv4/unicast   2          2    
                                                                     ipv6/unicast   2          2
```
